File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.cpp
#include <string>
#include <vector>
#include <cctype>
#include <cstdlib>
#include <algorithm>

using namespace std;

vector<char> prior{'+', '-', '*'};
vector<long long> nums;
vector<char> ops;

void init(string exp) {
    string tmp = "";
    for(int i = 0; i < exp.length(); i++) {
        if(isdigit(exp[i])) {
            tmp += exp[i];
        }
        else {
            ops.push_back(exp[i]);
            nums.push_back((long long)stoi(tmp));
            tmp = "";
        }
    }
    nums.push_back((long long)stoi(tmp));
}

long long cal(char op, long long a, long long b) {
    if(op == '+') return a + b;
    else if(op == '-') return a - b;
    else return a * b;
}

long long solution(string expression) {
    long long answer = 0;
    
    init(expression);
    sort(prior.begin(), prior.end());
    
    do {
        vector<long long> tmpNums = nums;
        vector<char> tmpOps = ops;
        
        for(char op: prior) {
            for(int idx = 0; idx < tmpOps.size(); idx++) {
                if(op == tmpOps[idx]) {
                    tmpNums[idx] = cal(op, tmpNums[idx], tmpNums[idx + 1]);
                    tmpNums.erase(tmpNums.begin() + idx + 1);
                    tmpOps.erase(tmpOps.begin() + idx);
                    idx--;
                }
            }
        }
        
        answer = max(answer, abs(tmpNums[0]));        
    } while(next_permutation(prior.begin(), prior.end()));
    
    return answer;
}

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
import re
from math import inf
from itertools import permutations

def calculate(nums: list, ops: list, priority: list) -> int:
    for p_op in priority:        
        while p_op in ops:
            idx = ops.index(p_op)
            if p_op == '*':
                n = nums[idx] * nums[idx + 1]
            elif p_op == '+':
                n = nums[idx] + nums[idx + 1]
            elif p_op == '-':
                n = nums[idx] - nums[idx + 1]
                
            nums[idx] = n
            del nums[idx + 1]
            del ops[idx]
    return abs(nums[0])

def solution(expression):
    num_pattern = '\d+'
    op_pattern = '[\*\+\-]'
    
    nums = list(map(int, re.findall(num_pattern, expression)))
    ops = re.findall(op_pattern, expression)
    
    answer = -inf
    tmp_priority = list(['+', '-', '*'])
    for priority in permutations(tmp_priority):
        result = calculate(nums.copy(), ops.copy(), priority)
        answer = max(answer, result)
    
    return answer